### PR TITLE
FIX-#4014: Apply deferred metadata before data egress

### LINKED ISSUE #4014: `df.to_csv` drops index column from CSV entirely
### System information
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**:
- **Modin version** (`modin.__version__`):
- **Python version**:
- **Code we can use to reproduce**:
```python
import modin.pandas as pd
cols = 100
arows = 20000
brows = 20000
keyrange = 100
a = pd.DataFrame(np.vstack([np.random.choice(keyrange,size=(arows)),np.random.normal(size=(cols,arows))]).transpose(),columns = ['key'] + ['avalue' + str(i) for i in range(1,1+cols)]).set_index('key')
b = pd.DataFrame(np.vstack([np.random.choice(keyrange,size=(brows)),np.random.normal(size=(cols,brows))]).transpose(),columns = ['key'] + ['bvalue' + str(i) for i in range(1,1+cols)]).set_index('key')
a.to_csv("a-big.csv")
b.to_csv("b-big.csv")
pd.read_csv("a-big.csv") # Has no key column
pd.read_csv("a-big.csv").set_index("key") # Errors
```
When using modin to write a Dataframe with a custom index to a csv file, the index is dropped entirely, and a range index is used when reading the Dataframe. Additionally, this may have been a transient error, but when I wrote the df using pandas, and then read it using modin, and called `set_index('key')` where `'key'` was the original index, the output displayed correctly, but I got the following `KeyError`:
```python
2022-01-20 13:23:49,267	ERROR worker.py:79 -- Unhandled error (suppress with RAY_IGNORE_UNHANDLED_ERRORS=1): ray::deploy_ray_func() (pid=78818, ip=127.0.0.1)
  File "/Users/rehandurrani/Documents/modin/modin/core/execution/ray/implementations/pandas_on_ray/partitioning/axis_partition.py", line 211, in deploy_ray_func
    result = func(*args)
  File "/Users/rehandurrani/Documents/modin/modin/core/dataframe/pandas/partitioning/axis_partition.py", line 160, in deploy_axis_func
    result = func(dataframe, **kwargs)
  File "/Users/rehandurrani/Documents/modin/modin/core/dataframe/pandas/dataframe/dataframe.py", line 1310, in _tree_reduce_func
    series_result = func(df, *args, **kwargs)
  File "/Users/rehandurrani/Documents/modin/modin/core/storage_formats/pandas/query_compiler.py", line 441, in map_func
    return pandas.merge(left, right, **kwargs)
  File "/Users/rehandurrani/.miniconda3/envs/modin/lib/python3.8/site-packages/pandas/core/reshape/merge.py", line 106, in merge
    op = _MergeOperation(
  File "/Users/rehandurrani/.miniconda3/envs/modin/lib/python3.8/site-packages/pandas/core/reshape/merge.py", line 699, in __init__
    ) = self._get_merge_keys()
  File "/Users/rehandurrani/.miniconda3/envs/modin/lib/python3.8/site-packages/pandas/core/reshape/merge.py", line 1096, in _get_merge_keys
    right_keys.append(right._get_label_or_level_values(rk))
  File "/Users/rehandurrani/.miniconda3/envs/modin/lib/python3.8/site-packages/pandas/core/generic.py", line 1779, in _get_label_or_level_values
    raise KeyError(key)
KeyError: 'key'
```

The code for the replication of this bug was borrowed from marsupialtail's [question](https://discuss.modin.org/t/modin-is-slower-than-pandas/255) on the [modin discourse](discuss.modin.org).

### PR DESCRIPTION
**Pull request by devin-petersohn**
Signed-off-by: Devin Petersohn <devin.petersohn@gmail.com>

## What do these changes do?

- [x] commit message follows format outlined [here](https://modin.readthedocs.io/en/latest/developer/contributing.html#commit-message-formatting)
- [x] passes `flake8 modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] passes `black --check modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] signed commit with `git commit -s` 
- [x] Resolves #4014  
- [x] tests added and passing
- [x] module layout described at `docs/development/architecture.rst` is up-to-date

### CODE REVIEW

**Review by naren-ponder — APPROVED** [review]
LGTM thanks Devin!

**Review by RehanSD — CHANGES_REQUESTED** [review]
This PR looks good to me, but I think there's a bug in the implementation of propagate index - this line 

https://github.com/modin-project/modin/blob/406af7c8d6c973d680ab9e07d6edc13eb69481f0/modin/core/dataframe/pandas/dataframe/dataframe.py#L534
Needs to be moved to the end of the function and be unindented to be outside of the if statement right? Cause otherwise we will error out whenever we try and propagate columns.

**Review by RehanSD — APPROVED** [review]
LGTM!

File: modin/core/execution/ray/implementations/pandas_on_ray/io/io.py [hunk 1]
```diff
@@ -111,6 +111,8 @@ def func(df):
             df.to_sql(**kwargs)
             return pandas.DataFrame()
 
+        # Ensure that the metadata is syncrhonized
+        qc._modin_frame._propagate_index_objs(axis=None)
```
Comment (YarShev) [inline modin/core/execution/ray/implementations/pandas_on_ray/io/io.py:115]: In all IO files where we need the metadata to be synchronized we use `synchronize_labels`. But here we call `_propagate_index_objs`. Should this be aligned?

**Review by YarShev — APPROVED** [review]
Left a question, but apart than that LGTM!

File: modin/core/execution/ray/implementations/pandas_on_ray/io/io.py (hunk 1 above)
Comment (devin-petersohn) [reply modin/core/execution/ray/implementations/pandas_on_ray/io/io.py:115]: `synchronize_labels` changes the deferred flag only, but `_propagate_index_objects` actually applies the metadata.

**Pull request merged**